### PR TITLE
Dedrickenc employee patient

### PR DESCRIPTION
**Pull request by DedrickEnc**
# Summary

This PR is refactoring the employee registration page and his test (integration and end to end)

# Details
here are some tasks performed : 

1. Update the database schema

2. Update  the server Interface

3. Refactor the user Interface

4. Add a registry for an employees

5. Clean the code

6. Integration test

7. End to End test

# More details

- The employee controller was fixed to create a debtor and a creditor properly

- The ```debtor_uuid``` reference was removed from the ```employee``` table 

- Now creating an **employee** will create a related **patient** directly  

- The UI interface for employee management was completely updated to follow the new standard

- **Employee** will have a card like a **patient** when he is created in the system

- Employee folder restructured as patient folder

- Some duplicated keys are removed from the translation files 

- The employee registry will keep ```update```, ```invoices```, ```payments``` 

- The old interface of employee for searching was removed and his test of integration 

- The all employee end-to-end test was skip now it is not

Thank you for contributing!

Before submitting this pull request, please verify that you have:
 - [x] Run your code through ESLint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
 - [x] Run integration tests.
 - [x] Run end-to-end tests.
 - [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

#1631

### CODE REVIEW

File: client/src/modules/employees/employee.service.js
Comment (jniles) [inline client/src/modules/employees/employee.service.js]: Just a note - `DepricatedFilterService` is name "deprecated" because new modules shouldn't be using it.  A much better filter service (`FilterService`) was introduced about a month ago in https://github.com/IMA-WorldHealth/bhima-2.X/pull/1566 that supports stronger filtering.

File: client/src/modules/employees/registration/employees.html
Comment (jniles) [inline client/src/modules/employees/registration/employees.html]: Please use the `translate` directive instead of the `translate` filter.  The repository has adopted this practice since https://github.com/IMA-WorldHealth/bhima-2.X/issues/1074.

File: client/src/modules/employees/registration/employees.html
Comment (jniles) [inline client/src/modules/employees/registration/employees.html:332]: Does an employee always have to have a service?  Where do guards and cleaning staff go?

File: client/src/modules/employees/registration/employees.html
Comment (jniles) [inline client/src/modules/employees/registration/employees.html:456]: What is the difference between `bank` and `bank_account`?  How are they meant to be used?

File: client/src/modules/employees/registration/employees.html
Comment (jniles) [inline client/src/modules/employees/registration/employees.html:492]: What application logic should happen if an employee is locked?  Can their patient entity still receive treatment?

File: client/src/modules/employees/registration/employees.js
Comment (jniles) [inline client/src/modules/employees/registration/employees.js]: Using ES2015+ is excellent!  However, we cannot use it on the client ... it breaks the production build.  To do this, we'll need to use a transpiler like [babeljs](https://babeljs.io/) or [TypeScript](https://www.typescriptlang.org/).

Please switch these back to ES5 until we have a transpiler included.

Here is the tracking issue: https://github.com/IMA-WorldHealth/bhima-2.X/issues/1439

File: client/src/modules/employees/registration/employees.html
Comment (jniles) [inline client/src/modules/employees/registration/employees.html:168]: Why does an employee need a creditor group?

File: client/src/modules/employees/registration/employees.js
Comment (jniles) [inline client/src/modules/employees/registration/employees.js]: I think this condition should already be caught in the `employeeForm.$invalid` case.  Are you seeing something different?

File: client/src/modules/employees/registry/registry.html
Comment (jniles) [inline client/src/modules/employees/registry/registry.html:4]: :+1: translate directive

File: client/src/modules/employees/registry/registry.js
Comment (jniles) [inline client/src/modules/employees/registry/registry.js]: It doesn't look like you actually use `GridSorting`.  Could you remove all imports that you don't need?

File: client/src/modules/employees/registry/registry.js
Comment (jniles) [inline client/src/modules/employees/registry/registry.js]: I don't think this column exists.

File: client/src/modules/employees/registry/registry.js
Comment (jniles) [inline client/src/modules/employees/registry/registry.js]: Do we need to see the `nb_enfant` by default?   It seems to me that some of these columns aren't needed on the main page and can be hidden by default with the option to expose them later.

File: client/src/modules/employees/registry/registry.js
Comment (jniles) [inline client/src/modules/employees/registry/registry.js]: In all other modules, these routes were merged into a single route to simplify the logic.  This was proposed in https://github.com/IMA-WorldHealth/bhima-2.X/issues/1665 and implemented in https://github.com/IMA-WorldHealth/bhima-2.X/pull/1674.

It shouldn't be too hard to update this PR before merge.

File: client/src/modules/employees/registry/registry.js
Comment (jniles) [inline client/src/modules/employees/registry/registry.js]: Since the employee card is just a patient card, why not call this `openPatientCard()`?

File: client/src/modules/employees/registry/search.modal.html
Comment (jniles) [inline client/src/modules/employees/registry/search.modal.html]: Out of curiosity, where did this style come from?

File: client/src/modules/employees/registry/search.modal.html
Comment (jniles) [inline client/src/modules/employees/registry/search.modal.html]: `bhFormDefaults` is deprecated.  Please do not use it in new modules.

Here is the tracking issue: https://github.com/IMA-WorldHealth/bhima-2.X/issues/1447

File: client/src/modules/employees/registry/search.modal.html
Comment (jniles) [inline client/src/modules/employees/registry/search.modal.html]: Please use the translate directive for all of these keys.

File: client/src/modules/employees/registry/search.modal.html
Comment (jniles) [inline client/src/modules/employees/registry/search.modal.html]: We have a component that performs this action.  It was described and implemented in https://github.com/IMA-WorldHealth/bhima-2.X/pull/1298.  Using it will clean up this code a lot.

File: client/src/modules/employees/registry/search.modal.html
Comment (jniles) [inline client/src/modules/employees/registry/search.modal.html]: Can you make this module `sex`?  It will line up with the `patient` table much better and keep everything in English.

File: client/src/modules/employees/registry/search.modal.html
Comment (jniles) [inline client/src/modules/employees/registry/search.modal.html]: This was tackled in https://github.com/IMA-WorldHealth/bhima-2.X/pull/1298.

File: client/src/modules/employees/registry/search.modal.js
Comment (jniles) [inline client/src/modules/employees/registry/search.modal.js]: These comments should probably be updated or removed.

File: client/src/modules/employees/templates/action.cell.html
Comment (jniles) [inline client/src/modules/employees/templates/action.cell.html]: These methods should refer to the action being performed.

File: server/controllers/finance/debtors/index.js
Comment (jniles) [inline server/controllers/finance/debtors/index.js]: Out of curiosity, why did this change to a LEFT JOIN?

File: server/controllers/medical/patients/index.js
Comment (jniles) [inline server/controllers/medical/patients/index.js:20]: :+1: Nice updates to comments.

File: server/models/schema.sql
Comment (jniles) [inline server/models/schema.sql:625]: Since most financial reports use the `debtor_uuid`, it makes sense to keep it in the table, I think.  That's why https://github.com/IMA-WorldHealth/bhima-2.X/issues/1631 only asked that that `patient_uuid` be added.

File: server/controllers/payroll/employees/index.js
Comment (jniles) [inline server/controllers/payroll/employees/index.js]: For best practice, set [autoParseStatements](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/lib/filter.js#L56) to `false`.  This option was added later to FilterParser (https://github.com/IMA-WorldHealth/bhima-2.X/commit/be9ebdcd68c22e0df178837d13baefd400b163d4) to ensure that the library didn't do anything the programmer had not explicitly requested.  It also helps the code read better.

File: test/end-to-end/employees/registration.page.js
Comment (jniles) [inline test/end-to-end/employees/registration.page.js]: This file has 22 lines of `page.X = X`.  In ES6, you can use a [class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes) to make the code read better.  Instead of this:
```js
function SomePage() {
  const page = this;

  setSomeProperty(value) {
    FU.input('$ctrl.key', value);
  }

  page.setSomeProperty = setSomeProperty;
}
```
... your code can be written as:
```js
class SomePage {
  setSomeProperty(value) {
    FU.input('$ctrl.key', value);
  }
}
```

In this case, the end behavior will be the same.

File: test/end-to-end/employees/registry.spec.js
Comment (jniles) [inline test/end-to-end/employees/registry.spec.js]: Just FYI:
```js
before(() => { return helpers.navigate(path)});
```
is the same as
```js
before(() => helpers.navigate(path));
```

**Review by jniles — CHANGES_REQUESTED** [review]
@DedrickEnc this represents an impressive amount of work.  Well done!

There are a number of parts of your code that are out of date.  I've commented on them to bring them up to date.  In general, here are the broad categories:
 1. Use the `translate` directive instead of the `translate` filter.
 2. Use the new `FilterService` instead of `DepricatedFilterService`.
 3. Use `bhClear` instead of `<span style="display: 'inline-block' " ...></span>` everywhere.
 4. Use ES5 on the client.

We've been using the `translate` directive for the last four months, you should change your code to use it here before this comes into master.  Also, using ES6 on the client will break the `gulp` build and prevent us from upgrading Vanga.  These should be fixed before this is merged into `master` so that we don't break our client's installation.

In the future, I think it would have been best to do one PR for creating patients when creating an employee, and a second PR for the employee registry.  It would be easier to rebase to master (since less code on each branch) and easier to review (since less code to read).

File: client/src/modules/employees/registration/employees.html
Comment (DedrickEnc) [reply client/src/modules/employees/registration/employees.html:168]: From the hospital perspective, a employee is a creditor so we need to know what will be his group of creditor because every creditor must have a group.

File: client/src/modules/employees/registration/employees.html
Comment (DedrickEnc) [reply client/src/modules/employees/registration/employees.html:332]: A service is not required for an employee in BHIMA, but in reality every employee belongs to a service. (for example at IMCK cleaners and guards belong to SERVICES GENERAUX ).

File: client/src/modules/employees/registration/employees.html
Comment (DedrickEnc) [reply client/src/modules/employees/registration/employees.html:456]: Bank is the name of the bank which is hosting the employee bank account, the bank account his a string of characters representing the employee bank account. There are not required

File: client/src/modules/employees/registration/employees.html
Comment (DedrickEnc) [reply client/src/modules/employees/registration/employees.html:492]: Honestly, I don't know. This should be discussed to help BHIMA find the right thing to do.

File: server/models/schema.sql
Comment (DedrickEnc) [reply server/models/schema.sql:625]: For the hospital, we have creditors and debtors. 
- Creditors are providers of services or drugs in this group we have (employees and suppliers)
- Debtors consumed services and drugs in this group we have (patients and other clients like visitors of guest house for the case of IMCK).
If you want we will discuss it, but I think it makes more sense to remove the ```debtor_uuid``` reference from the ```employee``` table. As the **employee** must have a correspondent **patient** that why there is a ```patient_uuid``` reference in the employee table.

File: client/src/modules/employees/registry/search.modal.html
Comment (DedrickEnc) [reply client/src/modules/employees/registry/search.modal.html]: Patient registry I think

File: server/controllers/finance/debtors/index.js
Comment (DedrickEnc) [reply server/controllers/finance/debtors/index.js]: Some debtors have not an entity_map so ```LEFT JOIN``` gave me every row I had also this problem with creditor and I did the same. I saw the triggers every thing seems well.

File: client/src/modules/employees/registry/search.modal.html
Comment (DedrickEnc) [reply client/src/modules/employees/registry/search.modal.html]: Thanks, there is an other directive to use in place?

Comment (DedrickEnc) [reply client/src/modules/employees/registry/search.modal.html]: Ok, I see

**Comment by jniles** [comment]
No, unfortunately.  In the linked issue, you will see that the Chrome
Browser actively prevents this type of directive from being created.  It
was a nice idea, but not really feasible at the moment without some serious
hacks.

File: server/controllers/finance/debtors/index.js
Comment (jniles) [reply server/controllers/finance/debtors/index.js]: If some debtors are missing an entity map, than that is a bug.  A trigger should create an an entity_map record anytime a debtor-type entity is created.

Do you have a reproducible test case?

File: server/models/schema.sql
Comment (jniles) [reply server/models/schema.sql:625]: Okay - the `patient_uuid` will still link to a `debtor`, so employees are still both debtors/creditors in BHIMA after this change.  I was just thinking that getting the `debtor_uuid` for an employee will now require that we do an extra JOIN on `patient`.

File: client/src/modules/employees/registry/search.modal.html
Comment (jniles) [reply client/src/modules/employees/registry/search.modal.html]: Normally when CSS is used in multiple locations, it is good practice to make it into a CSS `class`.  That way this element, and future elements, can benefit from sharing the same code base, the same bug fixes, and the whole application becomes smaller.

This is a non-blocking comment.

File: client/src/modules/employees/registration/employees.html
Comment (jniles) [reply client/src/modules/employees/registration/employees.html:332]: Okay, that makes sense.

We may need to separate the idea of services at Patient Invoice from these services.  We wouldn't want Patient Invoices to go to the SERVICES GENERAUX .. that data would be messy to clean up.

**Comment by DedrickEnc** [comment]
@jniles
Changes are landed

File: client/src/modules/employees/employee.service.js
Comment (jniles) [inline client/src/modules/employees/employee.service.js:4]: :+1:

File: client/src/modules/employees/employee.service.js
Comment (jniles) [inline client/src/modules/employees/employee.service.js:25]: In the future, it might be useful to have a default filter on `period` as well.

The feedback from Vanga is that they preferred the period selection over the date selection because it includes useful links like "yesterday"  and "last year".

File: client/src/modules/employees/registration/employees.html
Comment (jniles) [inline client/src/modules/employees/registration/employees.html]: At some point, we should remove this directive.  I'll make an issue about it for later.

**EDIT**: Issue added: #1739.

File: client/src/modules/employees/registration/employees.html
Comment (jniles) [inline client/src/modules/employees/registration/employees.html:119]: Could we add a link to where to add a grade here?  It might help users get set up faster.

File: client/src/modules/employees/registration/employees.html
Comment (jniles) [inline client/src/modules/employees/registration/employees.html]: Are you sure you want to `orderBy: 'text'` here, considering that you are only showing the `format` to the user?

File: client/src/modules/employees/registration/employees.html
Comment (jniles) [inline client/src/modules/employees/registration/employees.html:440]: It would probably be good to show the currency on this input, so that the users do not get confused.

File: client/src/modules/employees/registration/employees.js
Comment (jniles) [inline client/src/modules/employees/registration/employees.js]: This ES6 still needs to be cleaned up.

File: client/src/modules/employees/registry/registry.js
Comment (jniles) [inline client/src/modules/employees/registry/registry.js]: I don't think you used $translate.

File: client/src/modules/employees/registry/registry.js
Comment (jniles) [inline client/src/modules/employees/registry/registry.js:206]: 👍

File: client/src/modules/employees/registry/search.modal.js
Comment (jniles) [inline client/src/modules/employees/registry/search.modal.js]: This variable isn't ever used.

File: server/controllers/payroll/employees/index.js
Comment (jniles) [inline server/controllers/payroll/employees/index.js]: This should use `next` to handle the error.

**Review by jniles — COMMENTED** [review]
This is much better - the classes for page objects make much cleaner code, and the new filters are excellent. 👍   Good work.

The app still doesn't build with gulp-uglify to compress the files due to ES6 on the client.  I believe I found one place where an arrow function remains, there may be others.  You can build a minified version locally to test whether your code will run in production by changing the `isProduction` variable in our [gulpfile](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/gulpfile.js#L30)
```js
// old
const isProduction = (process.env.NODE_ENV === 'production');

// new
const isProduction = true;
```
This will help you catch any other errors here and in future pull requests.

There's also another spot where you could clear up your imports (`$translate`) since you don't use them.

After that, the code is good enough to be merged.

File: client/src/modules/employees/registration/employees.html
Comment (DedrickEnc) [reply client/src/modules/employees/registration/employees.html:119]: Yes but I think not for this PR I think. An other PR will follow for editing and other functionalities and I can implement it over there.
What do you think?

Comment (jniles) [reply client/src/modules/employees/registration/employees.html:119]: Sure - no problem.  Just a comment for the future.

**Comment by DedrickEnc** [comment]
@jniles
I have tried the build on production, every thing pass.

**Comment by DedrickEnc** [comment]
@jniles
I have tried the build on production, every thing pass.

**Comment by DedrickEnc** [comment]
All Es6 was removed from the client

**Review by jniles — APPROVED** [review]

**Comment by jniles** [comment]
@DedrickEnc, this PR passes code review.  All the comments have been responded to.  Well done :+1:

Since this is a large feature (2K+ line changes!), I propose that you present the module to the team on a projector Monday.  That way you can walk us through it, and the team can give feedback as we did with the report module last Thursday.

I'll defer to bors to merge it.

**Pull request merged**
